### ISSUE TITLE
Removing the "-93" switch when validating VHDL

### ISSUE THREAD

**Issue by justjustlife** [logisim-evolution/logisim-evolution#2098]
I successfully simulated a VHDL design using the process(all) construct, but validation failed, which is a significant issue. The failure during validation causes the corresponding circuit to be error-marked, even though the simulation runs without problems. Since the -93 switch is not used in simulation with vcom, it would be better to remove this switch from the validation process as well.

**Comment by justjustlife** [comment]
I initially thought the simulation had succeeded, but it seems I was mistaken. However, using process(all) worked successfully in both validation and simulation after making the following changes:
1. Source Code Modification:
  - Removed the -93 switch for validation.
  - Updated the modelsim.ini file included in the source code to use VHDL-2008 instead of VHDL-2002.
2. Local Installation Update:
  - Modified the modelsim.ini file in my local ModelSim installation, changing the VHDL setting from VHDL-2002 to VHDL-2008.

It seems the current behavior is as follows:
1. Validation:
  - Uses the local modelsim.ini file but is overridden by the -93 switch.
2. Simulation:
  - Uses the modelsim.ini file included in the source code.

Suggested Improvements
1. Remove the -93 switch entirely for validation.
2. Configure validation to use the modelsim.ini file included in the source code instead of the local installation file.
3. Update the modelsim.ini file in the source code to default to VHDL-2008.

These changes would ensure consistent behavior between validation and simulation while aligning both with the VHDL-2008 standard.

**Comment by hewzhew** [comment]
I checked the current `main` branch again. The validation path now copies the bundled `modelsim.ini` into the temporary validation directory, and the generated `vcom` validation command no longer contains an explicit `-93` option. That was addressed by #2642.

So I think the original mismatch reported here, validation using a different configuration than simulation, should now be fixed. The remaining idea from the discussion, making the VHDL language version selectable, seems like a broader enhancement and may deserve a separate issue/design discussion if it is still wanted.

Would it make sense to close this issue, or keep it open specifically for the selectable VHDL standard work?

**Comment by maehne** [comment]
I think it is best to keep this issue open, as the initial motivation for this issue clearly was the wish to use VHDL-2008 constructs in Logisim-evolution. So, a new PR implementing the selection of the VHDL language version should reference again this issue.

**Comment by hewzhew** [comment]
I did a first pass through the current code paths for this issue.

My understanding is:

- The original validation/simulation mismatch is now mostly addressed by #2642: validation no longer passes an explicit `-93`, and it copies the bundled `modelsim.ini` into the temporary validation directory.
- The remaining enhancement is to make the VHDL language version selectable, instead of changing the bundled default from VHDL-2002 to VHDL-2008.

A possible implementation direction would be:

1. Add a project/file-level VHDL language version option saved in the `.circ` file, defaulting to VHDL-2002 for compatibility with existing projects and toolchains.
2. Offer at least VHDL-1993, VHDL-2002, and VHDL-2008.
3. Keep validation and simulation consistent by using a shared helper that copies the bundled `modelsim.ini` and rewrites only its `VHDL93 = ...` setting in the temporary validation/simulation directory.
4. Add unit coverage for the config rewriting, while keeping the existing check that validation does not force `-93` on the command line.

I currently think project-level is preferable to a global application preference because the VHDL source is stored in the `.circ` file, so the required language version should travel with the project. Does this sound like the right direction before I start preparing a PR?

**Comment by maehne** [comment]
A good overview on the differences of the different VHDL standards is provided in the [GHDL documentation](https://ghdl.github.io/ghdl/using/ImplementationOfVHDL.html). As of this writing, 4 standard groups notably differ from each other: 1987, 1993, 2008, and 2019. Support for 2008 and 2019 is improving in the open-source simulators GHDL and nvc. Unfortunately, there are still FPGA toolchains, which at least in the free tier don't support VHDL-2008. Support for the different VHDL versions varies also in the commercial simulators and evolves from release to release. (A simular situation holds for the different Verilog/SystemVerilog standards.) For the moment, we only support ModelSIM/Questa Sim, but that may change in the future, as there is interest for better support for the fully open-source tool chains.

For the moment, our VHDL/Verilog net listing does not make any use of VHDL-2008/SystemVerilog features, respectively. Therefore, the issue currently mostly affects sub-circuits described directly as VHDL design entities.

For the moment, I would make the VHDL standard (and maybe in the future Verilog/SystemVerilog standard) selectable on the Software or FPGA Commander tab of the application Preferences. The selected VHDL version shall be passed as option to the configured simulation/synthesis tool. I think it is out of scope for our tool to keep track of vendor support for different HDL versions, i.e., it is fine if a tool rejects the passed HDL version and errors out. IMHO, it is more visible to the user to select the VHDL version in the app settings and maybe later the VHDL Entity properties than in the project's preferences. Some tools allow mixing HDL code conformant to different standards for joint simulation or synthesis. In that case, we may later consider to add an HDL version attribute to VHDL Entity's properties. Other tools may insist that all HDL files are conformant to a single common HDL version. 

IMHO, we should keep it simple and stick for the moment to an application setting, which is used for analysis of all VHDL files. Once we see demand, we may add the HDL version as attribute to VHDL Entities created by the user.

@BFH-ktt1, @davidhutchens, @kevinawalsh: What's your take on this?

**Comment by hewzhew** [comment]
Thanks for the clarification. I rechecked the current code paths and my understanding is that the original validation/simulation mismatch is mostly addressed now: validation no longer passes an explicit `-93`, and both validation and simulation copy the bundled `modelsim.ini` into their temporary working directories.

So the remaining scope of this issue seems to be the selectable VHDL language version.

A small first step could be:

- add an application-level preference named something like `VHDL standard`, defaulting to the current VHDL-2002 behavior;
- initially offer VHDL-1993, VHDL-2002, and VHDL-2008;
- place the option on the Software tab, close to the QuestaSim/ModelSim validation/path settings, since it directly controls how the external VHDL tool compiles user VHDL entities;
- keep validation and simulation consistent by using a shared helper that copies the bundled `modelsim.ini` and rewrites only the `[vcom]` `VHDL93 = ...` entry in the temporary directory;
- avoid changing the bundled default `modelsim.ini` itself.

This would preserve the current default behavior for existing users and toolchains, while letting users who need VHDL-2008 constructs such as `process(all)` opt into that mode explicitly.

If this scope sounds reasonable, I can prepare a focused PR for this first application-level preference. Per-entity VHDL standard settings and newer standards such as VHDL-2019 could remain follow-up work if there is later demand.

**Comment by maehne** [comment]
@hewzhew: I agree with the suggested approach. Further improvements in this area should be done once users explicitly express interest in this area.
